AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/InjectHelper.java

```diff
@@ -0,0 +1,152 @@
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.VariableElement;
+
+import org.androidannotations.ElementValidation;
+import org.androidannotations.annotations.NonConfigurationInstance;
+import org.androidannotations.holder.EComponentHolder;
+import org.androidannotations.holder.HasMethodInjection;
+
+import com.sun.codemodel.JBlock;
+import com.sun.codemodel.JExpression;
+import com.sun.codemodel.JFieldRef;
+import com.sun.codemodel.JInvocation;
+
+public class InjectHelper {
+
+	private final Map<ExecutableElement, List<ParamHelper>> methodParameterMap = new HashMap<>();
+
+	private final ValidatorHelper validatorHelper;
+	private final HasMethodInjection handler;
+
+	public InjectHelper(ValidatorHelper validatorHelper, HasMethodInjection handler) {
+		this.validatorHelper = validatorHelper;
+		this.handler = handler;
+	}
+
+	public void validate(Class<? extends Annotation> expectedAnnotation, Element element, ElementValidation valid) {
+		Element enclosingElement = element.getEnclosingElement();
+		if (element instanceof VariableElement && enclosingElement instanceof ExecutableElement) {
+			validatorHelper.enclosingElementHasEnhancedComponentAnnotation(enclosingElement, valid);
+			validatorHelper.param.annotatedWith(expectedAnnotation).multiple().validate((ExecutableElement) enclosingElement, valid);
+		} else {
+			validatorHelper.enclosingElementHasEnhancedComponentAnnotation(element, valid);
+		}
+	}
+
+	public void process(Element element, EComponentHolder holder) {
+		if (element instanceof ExecutableElement) {
+			processMethod(element, holder);
+		} else {
+			Element enclosingElement = element.getEnclosingElement();
+			if (enclosingElement instanceof ExecutableElement) {
+				processParam(element, holder);
+			} else {
+				processField(element, holder);
+			}
+		}
+	}
+
+	private void processParam(Element element, EComponentHolder holder) {
+		ExecutableElement method = (ExecutableElement) element.getEnclosingElement();
+		List<? extends VariableElement> parameters = method.getParameters();
+		List<ParamHelper> paramHelper = methodParameterMap.get(method);
+		int paramCount = parameters.size();
+
+		if (paramHelper == null) {
+			paramHelper = new ArrayList<>();
+			methodParameterMap.put(method, paramHelper);
+		}
+
+		for (int paramIndex = 0; paramIndex < paramCount; paramIndex++) {
+			VariableElement param = parameters.get(paramIndex);
+			if (param.equals(element)) {
+				JExpression getInstance = handler.getInstanceInvocation(param, holder, param);
+				paramHelper.add(new ParamHelper(getInstance, paramIndex));
+			}
+		}
+
+		if (paramHelper.size() == paramCount) {
+			String methodName = method.getSimpleName().toString();
+
+			Collections.sort(paramHelper, new Comparator<ParamHelper>() {
```

Comment:
Instead of the anonymous `Comparator` class, it would be nicer to have `ParamHelper` implement `Comparable`.